### ISSUE TITLE
[Topic creation inside the same namespace where Kafka 2.4.0 and 2.6.0 are running] ...

### ISSUE THREAD

**Issue by vperi1730** [strimzi/strimzi-kafka-operator#3942]
Hi Team,

I need clarification on one of the use case which we are testing in our GKE clusters. 

Under a single namespace, I have 2 versions of Strimzi(17 and 20) with Kafka(2.4.0 and 2.6.0) running. Now I would like to create a simple Kafka topic in 2.4.0 for verifying the producing and consuming scenarios. 

Ask is when the topic is created inside the 2.4.0 application, will it also be auto-created inside the other Kafka of 2.6.0 where the Strimzi operator 20 would come into the play. If not what would be the behavior?

It looks something like below.

This is for 17 and 2.4.0
```
pod/cluster17-kafka-0                             1/1     Running   0          5h
pod/cluster17-kafka-1                             1/1     Running   0          101m
pod/cluster17-kafka-2                             1/1     Running   0          52m
pod/cluster17-zookeeper-0                         1/1     Running   0          5h7m
pod/cluster17-zookeeper-1                         1/1     Running   0          5h7m
pod/cluster17-zookeeper-2                         1/1     Running   0          5h7m
pod/strimzi-cluster-operator17-58c487f694-6746g   1/1     Running   0          5h11m
```

This is for 20 and 2.6.0
```
pod/cluster20-kafka-0                             1/1     Running   0          5h
pod/cluster20-kafka-1                             1/1     Running   0          101m
pod/cluster20-kafka-2                             1/1     Running   0          52m
pod/cluster20-zookeeper-0                         1/1     Running   0          5h7m
pod/cluster20-zookeeper-1                         1/1     Running   0          5h7m
pod/cluster20-zookeeper-2                         1/1     Running   0          5h7m
pod/strimzi-cluster-operator20-58c487f694-6746g   1/1     Running   0          5h11m
```

Need your inputs on this.

**Comment by scholzj** [comment]
The general advice is to not run multiple Kafka clusters in the same namespace because it creates exactly the issues like this. Maybe @tombentley or @sknot-rh would know more about the exact TO behaviour. But for example with the User operator, you would be able to have user with specific name only in one of these clusters.

**Comment by vperi1730** [comment]
I agree with your point, However, if I want to maintain both the Kafka versions inside the same namespace, what approach would you take or what would you recommend?

**Comment by scholzj** [comment]
In general, each resource can exist only once under given name in given namespace. The `strimzi.io/cluster` defines which cluster will use the resource. So KafkaUser names `user1` with the label pointing to `cluster-a` will be created in cluster A. And if you want to create the the `user1` also in cluster B, you cannot do it.

Topics are more complicated because they have the name in the spec and they are also created by the Topic Operator. So I have no idea how that would behave - that is what the other would need to explain.

Another issue is that you cannot have easily two operators watch the same namespace. So I do not think you will be able to have one cluster managed with Strimzi 0.17 and one with 0.20 in the same namespace. 

So as I said ... just don't do this :-o.

**Comment by vperi1730** [comment]
I agree with you on this, Let me try certain use cases to self answer to myself being a technical person just to see how things behave. Thanks for the information :)

Another question for the below scenario.

1) Launched 17 operator deployment.
2) Launched Kafka 2.4.0
3) Launched 20 operator deployment. 

In this flow, there is no issue with Kafka as an application as the deployment went successfully. However, in the Strimzi 20 operator logs, i see the following logs which tell me that the operator is looking for the Kafka application which is running inside that namespace. As 2.4.0 is something un-supported it is throwing invalid-resource exception.

```
Caused by: io.strimzi.operator.cluster.model.NoImageException: No image for version 2.4.0 in {2.5.1=strimzi/kafka:latest-kafka-2.5.1, 2.6.0=strimzi/kafka:latest-kafka-2.6.0, 2.5.0=strimzi/kafka:latest-kafka-2.5.0}
        at io.strimzi.operator.cluster.model.KafkaVersion$Lookup.image(KafkaVersion.java:163) ~[io.strimzi.cluster-operator-0.21.0-SNAPSHOT.jar:0.21.0-SNAPSHOT]
        at io.strimzi.operator.cluster.model.KafkaVersion$Lookup.kafkaImage(KafkaVersion.java:182) ~[io.strimzi.cluster-operator-0.21.0-SNAPSHOT.jar:0.21.0-SNAPSHOT]
        ... 18 more
2020-11-12 03:40:10 WARN  AbstractOperator:470 - Reconciliation #126(watch) Kafka(kafka20/cluster17): Failed to reconcile
io.strimzi.operator.cluster.model.InvalidResourceException: Version 2.4.0 is not supported. Supported versions are: 2.5.0, 2.5.1, 2.6.0.
```

The question here is how do I tell my operator to watch only its supported version in the namespace by ignoring the rest which is not instead of shooting back exceptions? Is this achievable/possible?

**Comment by scholzj** [comment]
As I said, you cannot have two operators watch the same namespace. The error with the unsupported version is one of the way how it shows up. You can try to use different namespaces in the same cluster. But it has some things to be careful about:
* You need to always use the latest CRDs and ClusterRoles
* You will need to rename the ClusterRoleBindings when installing them otherwise they would rewrite each other.

**Comment by vperi1730** [comment]
Ok, I believe it should also be good enough if i maintain only Strmizi version 20 CRD's, ClusterRoles, and ClusterRolebindings inside my cluster which should work for both Kafka's running in 2 diff namespaces(2.4.0 and 2.6.0). Do you agree with this?

**Comment by scholzj** [comment]
That is what I said for CRDs and for Cluster Roles. But the ClusterRoleBindings are different. They give the rights to a specific service account in a specific namespace. So you cannot use just one. You need to rename them and have them twice -> one set for each SA/namespace.

**Comment by vperi1730** [comment]
Perfect, Makes sense Scholzj :),  Thanks for it.
 
Any thoughts on Q3 above?

**Comment by vperi1730** [comment]
During my testing, i have also launched MM2 for which i started noticing Kafka connect is getting stopped. Can you help me with this?

```
2020-11-12 13:21:24,145 WARN [Consumer clientId=consumer-mirrormaker2-cluster-1, groupId=mirrormaker2-cluster] 10 partitions have leader brokers without a matching listener, including [mirrormaker2-cluster-offsets-11, mirrormaker2-cluster-offsets-16, mirrormaker2-cluster-offsets-3, mirrormaker2-cluster-offsets-23, mirrormaker2-cluster-offsets-13, mirrormaker2-cluster-offsets-18, mirrormaker2-cluster-offsets-8, mirrormaker2-cluster-offsets-1, mirrormaker2-cluster-offsets-6, mirrormaker2-cluster-offsets-21] (org.apache.kafka.clients.NetworkClient) [DistributedHerder-connect-1-1]
2020-11-12 13:21:24,247 WARN [Consumer clientId=consumer-mirrormaker2-cluster-1, groupId=mirrormaker2-cluster] 10 partitions have leader brokers without a matching listener, including [mirrormaker2-cluster-offsets-11, mirrormaker2-cluster-offsets-16, mirrormaker2-cluster-offsets-3, mirrormaker2-cluster-offsets-23, mirrormaker2-cluster-offsets-13, mirrormaker2-cluster-offsets-18, mirrormaker2-cluster-offsets-8, mirrormaker2-cluster-offsets-1, mirrormaker2-cluster-offsets-6, mirrormaker2-cluster-offsets-21] (org.apache.kafka.clients.NetworkClient) [DistributedHerder-connect-1-1]
2020-11-12 13:21:24,311 ERROR [Worker clientId=connect-1, groupId=mirrormaker2-cluster] Uncaught exception in herder work thread, exiting:  (org.apache.kafka.connect.runtime.distributed.DistributedHerder) [DistributedHerder-connect-1-1]
org.apache.kafka.common.errors.TimeoutException: Failed to get offsets by times in 30001ms
2020-11-12 13:21:24,323 INFO Kafka Connect stopping (org.apache.kafka.connect.runtime.Connect) [connect-shutdown-hook]
2020-11-12 13:21:24,325 INFO Stopping REST server (org.apache.kafka.connect.runtime.rest.RestServer) [connect-shutdown-hook]
2020-11-12 13:21:24,336 INFO Stopped http_8083@7e8a46b7{HTTP/1.1,[http/1.1]}{0.0.0.0:8083} (org.eclipse.jetty.server.AbstractConnector) [Thread-1]
2020-11-12 13:21:24,340 INFO node0 Stopped scavenging (org.eclipse.jetty.server.session) [Thread-1]
2020-11-12 13:21:24,346 INFO REST server stopped (org.apache.kafka.connect.runtime.rest.RestServer) [connect-shutdown-hook]
2020-11-12 13:21:24,347 INFO [Worker clientId=connect-1, groupId=mirrormaker2-cluster] Herder stopping (org.apache.kafka.connect.runtime.distributed.DistributedHerder) [connect-shutdown-hook]
2020-11-12 13:21:29,349 INFO [Worker clientId=connect-1, groupId=mirrormaker2-cluster] Herder stopped (org.apache.kafka.connect.runtime.distributed.DistributedHerder) [connect-shutdown-hook]
2020-11-12 13:21:29,349 INFO Kafka Connect stopped (org.apache.kafka.connect.runtime.Connect) [connect-shutdown-hook]
```
MM2 CR.

```
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaMirrorMaker2
metadata:
  name: mirror-maker-20
spec:
  version: 2.6.0
  replicas: 1
  connectCluster: "cluster20"
  clusters:
  - alias: "cluster17"
    bootstrapServers: cluster17-kafka-bootstrap.kafka20.svc:9092
  - alias: "cluster20"
    bootstrapServers: cluster20-kafka-bootstrap.kafka20t.svc:9092
    config:
      config.storage.replication.factor: 1
      offset.storage.replication.factor: 1
      status.storage.replication.factor: 1
  mirrors:
  - sourceCluster: "cluster17"
    targetCluster: "cluster20"
    sourceConnector:
      config:
        replication.factor: 1
        offset-syncs.topic.replication.factor: 1
        sync.topic.acls.enabled: "false"
    heartbeatConnector:
      config:
        heartbeats.topic.replication.factor: 1
    checkpointConnector:
      config:
        checkpoints.topic.replication.factor: 1
    topicsPattern: ".*"
    groupsPattern: ".*"

```

**Comment by vperi1730** [comment]
Also while launching the MM2, i am noticing of Kafka connect getting stopped. Here are some of the logs from it.

```
2020-11-12 13:31:25,430 WARN [Consumer clientId=consumer-mirrormaker2-cluster-1, groupId=mirrormaker2-cluster] 10 partitions have leader brokers without a matching listener, including [mirrormaker2-cluster-offsets-11, mirrormaker2-cluster-offsets-16, mirrormaker2-cluster-offsets-3, mirrormaker2-cluster-offsets-23, mirrormaker2-cluster-offsets-13, mirrormaker2-cluster-offsets-18, mirrormaker2-cluster-offsets-8, mirrormaker2-cluster-offsets-1, mirrormaker2-cluster-offsets-6, mirrormaker2-cluster-offsets-21] (org.apache.kafka.clients.NetworkClient) [DistributedHerder-connect-1-1]
2020-11-12 13:31:25,532 WARN [Consumer clientId=consumer-mirrormaker2-cluster-1, groupId=mirrormaker2-cluster] 10 partitions have leader brokers without a matching listener, including [mirrormaker2-cluster-offsets-11, mirrormaker2-cluster-offsets-16, mirrormaker2-cluster-offsets-3, mirrormaker2-cluster-offsets-23, mirrormaker2-cluster-offsets-13, mirrormaker2-cluster-offsets-18, mirrormaker2-cluster-offsets-8, mirrormaker2-cluster-offsets-1, mirrormaker2-cluster-offsets-6, mirrormaker2-cluster-offsets-21] (org.apache.kafka.clients.NetworkClient) [DistributedHerder-connect-1-1]
2020-11-12 13:31:25,634 WARN [Consumer clientId=consumer-mirrormaker2-cluster-1, groupId=mirrormaker2-cluster] 10 partitions have leader brokers without a matching listener, including [mirrormaker2-cluster-offsets-11, mirrormaker2-cluster-offsets-16, mirrormaker2-cluster-offsets-3, mirrormaker2-cluster-offsets-23, mirrormaker2-cluster-offsets-13, mirrormaker2-cluster-offsets-18, mirrormaker2-cluster-offsets-8, mirrormaker2-cluster-offsets-1, mirrormaker2-cluster-offsets-6, mirrormaker2-cluster-offsets-21] (org.apache.kafka.clients.NetworkClient) [DistributedHerder-connect-1-1]
2020-11-12 13:31:25,708 ERROR [Worker clientId=connect-1, groupId=mirrormaker2-cluster] Uncaught exception in herder work thread, exiting:  (org.apache.kafka.connect.runtime.distributed.DistributedHerder) [DistributedHerder-connect-1-1]
org.apache.kafka.common.errors.TimeoutException: Failed to get offsets by times in 30000ms
2020-11-12 13:31:25,732 INFO Kafka Connect stopping (org.apache.kafka.connect.runtime.Connect) [connect-shutdown-hook]
2020-11-12 13:31:25,749 INFO Stopping REST server (org.apache.kafka.connect.runtime.rest.RestServer) [connect-shutdown-hook]
2020-11-12 13:31:25,763 INFO Stopped http_8083@7e8a46b7{HTTP/1.1,[http/1.1]}{0.0.0.0:8083} (org.eclipse.jetty.server.AbstractConnector) [Thread-1]
2020-11-12 13:31:25,764 INFO node0 Stopped scavenging (org.eclipse.jetty.server.session) [Thread-1]
2020-11-12 13:31:25,771 INFO REST server stopped (org.apache.kafka.connect.runtime.rest.RestServer) [connect-shutdown-hook]
2020-11-12 13:31:25,771 INFO [Worker clientId=connect-1, groupId=mirrormaker2-cluster] Herder stopping (org.apache.kafka.connect.runtime.distributed.DistributedHerder) [connect-shutdown-hook]
2020-11-12 13:31:30,777 INFO [Worker clientId=connect-1, groupId=mirrormaker2-cluster] Herder stopped (org.apache.kafka.connect.runtime.distributed.DistributedHerder) [connect-shutdown-hook]
2020-11-12 13:31:30,777 INFO Kafka Connect stopped (org.apache.kafka.connect.runtime.Connect) [connect-shutdown-hook]
```

Can someone please help me here?

**Comment by scholzj** [comment]
> Any thoughts on Q3 above? - My understanding from previous executions is that ideally, we should have only 1 entity operator pod in running state.

Re Q3: To be honest, not sure what would cause this. I guess the eviction would happen by Kubernetes (Strimzi really creates the Deployment. The rest is driven by Kubernetes it self.) But in theory running two operators watching the same namespace can cause similar things when they fight over the control of the Kafka clusters.

As for the MM2, I have no idea ... it looks like an issue when connecting to Kafka to me.
